### PR TITLE
Backport cling fixes for CMS

### PR DESCRIPTION
**Pull request by vepadulano**
Backports of #14287 and #14352

### CODE REVIEW

**Review by dpiparo — APPROVED** [review]
if tests pass, we can merge.

**Comment by vepadulano** [comment]
Unfortunately the commits do not even compile with some scary error
```
 /home/sftnight/build/workspace/root-pullrequests-build/root/core/metacling/src/TClingCallbacks.cxx: In constructor ‘AutoloadLibraryMU::AutoloadLibraryMU(const TClingCallbacks&, const string&, const SymbolNameVector&)’:
/home/sftnight/build/workspace/root-pullrequests-build/root/core/metacling/src/TClingCallbacks.cxx:90:120: error: cannot allocate an object of abstract type ‘llvm::orc::MaterializationUnit’
   90 |       : MaterializationUnit({getSymbolFlagsMap(Symbols), nullptr}), fCallbacks(cb), fLibrary(Library), fSymbols(Symbols)
```

Can this be some difference between LLVM 13 and 16 that's causing it? @vgvassilev

File: core/metacling/src/TClingCallbacks.cxx [hunk 1]
```diff
@@ -82,16 +82,24 @@ extern "C" {
...
 
 class AutoloadLibraryMU : public llvm::orc::MaterializationUnit {
+   const TClingCallbacks &fCallbacks;
+   std::string fLibrary;
+   llvm::orc::SymbolNameVector fSymbols;
 public:
-   AutoloadLibraryMU(const std::string &Library, const llvm::orc::SymbolNameVector &Symbols)
-      : MaterializationUnit(getSymbolFlagsMap(Symbols), nullptr), fLibrary(Library), fSymbols(Symbols)
+   AutoloadLibraryMU(const TClingCallbacks &cb, const std::string &Library, const llvm::orc::SymbolNameVector &Symbols)
+      : MaterializationUnit({getSymbolFlagsMap(Symbols), nullptr}), fCallbacks(cb), fLibrary(Library), fSymbols(Symbols)
```
Comment (vgvassilev) [inline core/metacling/src/TClingCallbacks.cxx:90]:
```suggestion
      : MaterializationUnit(getSymbolFlagsMap(Symbols), nullptr), fCallbacks(cb), fLibrary(Library), fSymbols(Symbols)
```

How about that one.

Comment (vepadulano) [reply core/metacling/src/TClingCallbacks.cxx:90]: Let me try

Comment (vepadulano) [reply core/metacling/src/TClingCallbacks.cxx]: The build finishes fine on my machine, I have updated the commit

**Comment by smuzaffar** [comment]
thanks for the backport, I have started CMSSW tests here https://github.com/cms-sw/root/pull/195

**Comment by vepadulano** [comment]
> thanks for the backport, I have started CMSSW tests here https://github.com/cms-sw/root/pull/195

Hi @smuzaffar , thanks for your promptness! We discovered there was a small wrinkle preventing proper compilation in 6.30, I have uploaded an updated commit which should work properly literally a few minutes ago. I just wanted to make sure that the PR on CMSSW takes this latest version of the commits

**Comment by vepadulano** [comment]
Ok now it compiles but we get the following error in the new test
```cpp
  input_line_49:1:21: error: unknown type name 'ROOT_7459'
  int ROOT_7459 = 42; ROOT_7459++;
                      ^
  input_line_49:1:30: error: expected unqualified-id
  int ROOT_7459 = 42; ROOT_7459++;
```

**Comment by dpiparo** [comment]
I would like to point out that this is a backport of a very generic fix, which works very well with llvm16, which is not in the branch 6.30. A solution of the issue reported by CMS was already made available in December: if you want, this is a more refined version of that.

**Comment by vgvassilev** [comment]
> Ok now it compiles but we get the following error in the new test
> 
> ```c++
>   input_line_49:1:21: error: unknown type name 'ROOT_7459'
>   int ROOT_7459 = 42; ROOT_7459++;
>                       ^
>   input_line_49:1:30: error: expected unqualified-id
>   int ROOT_7459 = 42; ROOT_7459++;
> ```

Somehow the parsing seems broken. We should probably try using `ProcessLine` instead of `Declare`... still a mystery why this fails only for the PCH based builds...

**Comment by vepadulano** [comment]
> We should probably try using ProcessLine instead of Declare

Seems like an easy test, let me try that :+1:

**Comment by vepadulano** [comment]
At least on my machine substituting `Declare` with `ProcessLine` fixed the test failure, let's see what the CI has to say

**Comment by vgvassilev** [comment]
> At least on my machine substituting `Declare` with `ProcessLine` fixed the test failure, let's see what the CI has to say

We need to land that change in the master for consistency reasons.

**Comment by vepadulano** [comment]
@vgvassilev The test now passes on all platforms. Let me know what you think about the commit and if we can merge this PR

**Review by vgvassilev — APPROVED** [review]
LGTM!

**Comment by smuzaffar** [comment]
@dpiparo , thanks for the backport. CMSSW tests via https://github.com/cms-sw/root/pull/195 for this change look good. I just have opened https://github.com/cms-sw/cmsdist/pull/8949 to integrate latest v6-30-00-patches changes in to CMSSW 14.0.X

**Pull request merged**